corehq/apps/user_importer/importer.py

```diff
@@ -578,3 +580,127 @@ def _get_domain_info(domain):
+            first_name = row.get('first name')
+            last_name = row.get('last name')
+            role = row.get('role', None)
+            status = row.get('status')
+
+            try:
+                remove = spec_value_to_boolean_or_none(row, 'remove')
+                role_qualified_id = None
+
+                if role:
+                    role_qualified_id = roles_by_name[role].get_qualified_id()
+                if email:
+                    email = email.lower()
+                if username:
+                    if not upload_user.can_edit_web_users():
+                        raise UserUploadError(_(
+                            "Only users with the edit web users permission can upload web users"
+                        ))
+
+                    user = CouchUser.get_by_username(username)
+                    if user:
+                        if remove:
+                            if not user or not user.is_member_of(domain):
+                                raise UserUploadError(_(
+                                    "You cannot remove a web user that is not a member of this project. "
+                                    "{username} is not a member.").format(username=username)
+                                )
+                            else:
+                                user.delete_domain_membership(domain)
+                                user.save()
+                        else:
+                            if not role:
+                                raise UserUploadError(_(
+                                    "You cannot upload a web user without a role. {username} does not have "
+                                    "a role").format(username=username)
+                                )
+
+                            if user.is_member_of(domain):
+                                user_current_role = user.get_role(domain=domain)
+                                role_updated = not (user_current_role
+                                                    and user_current_role.get_qualified_id() == role_qualified_id)
+                                if role_updated:
+                                    user.set_role(domain, role_qualified_id)
+                                if first_name:
+                                    user.first_name = first_name
+                                if last_name:
+                                    user.last_name = last_name
+                                status_row['flag'] = 'updated'
+                                user.save()
+
+                            elif not user.is_member_of(domain) and status == 'Active User':
+                                user.add_as_web_user(domain, role=role_qualified_id)
+                                status_row['flag'] = 'updated'
+
+                            elif not user.is_member_of(domain) and status == 'Invited':
+                                invite = Invitation.objects.filter(email=email, domain=domain).first()
```

Comment:
It is a little bit confusing that invites work on the email field rather than the username one, since it becomes the username. Username should also be a required field, and that makes this easier, so we will know they exist, and ensures consistency across user types.